src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ExceptionsStoreTests.cs

```diff
@@ -0,0 +1,385 @@
+            {
+                AddExceptionInstance(store, cache, ExpectedId3);
+
+                await thresholdCallback3.WaitForThresholdsAsync(CommonTestTimeouts.GeneralTimeout);
+            }
+
+            ThresholdCallback thresholdCallback4 = new(0, 2);
+            using (CallbackRegistration.Register(aggregateCallback, thresholdCallback4))
+            {
+                RemoveExceptionInstance(store, ExpectedId2);
+                RemoveExceptionInstance(store, ExpectedId3);
+
+                await thresholdCallback4.WaitForThresholdsAsync(CommonTestTimeouts.GeneralTimeout);
+            }
+
+            // Assert
+            Assert.Empty(store.GetSnapshot());
+        }
+
+        private static IExceptionsNameCache CreateCache()
+        {
+            EventExceptionsPipelineNameCache cache = new();
+            cache.AddExceptionGroup(DefaultExceptionGroupId, 1, 1, 0);
+            return cache;
+        }
+
+        private static void AddExceptionInstance(ExceptionsStore store, IExceptionsNameCache cache, ulong exceptionId)
+        {
+            store.AddExceptionInstance(
+                cache,
+                exceptionId,
+                DefaultExceptionGroupId,
+                null,
+                DateTime.UtcNow,
+                Array.Empty<ulong>(),
+                0,
+                Array.Empty<ulong>(),
+                null,
+                ActivityIdFormat.Unknown);
+        }
+
+        private static void RemoveExceptionInstance(ExceptionsStore store, ulong exceptionId)
+        {
+            store.RemoveExceptionInstance(exceptionId);
+        }
+
+        private sealed class PassThroughCallbackFactory : IExceptionsStoreCallbackFactory
+        {
+            private readonly IExceptionsStoreCallback _callback;
+
+            public PassThroughCallbackFactory(IExceptionsStoreCallback callback)
+            {
+                _callback = callback ?? throw new ArgumentNullException(nameof(callback));
+            }
+
+            public IExceptionsStoreCallback Create(IExceptionsStore store)
+            {
+                return _callback;
+            }
+        }
+
+        private sealed class ThresholdCallback : ExceptionsStoreCallbackBase
+        {
+            private readonly TaskCompletionSource _addCompletionSource =
+                new(TaskCreationOptions.RunContinuationsAsynchronously);
+            private readonly int _addThreshold;
+
+            private readonly TaskCompletionSource _removeCompletionSource =
+                new(TaskCreationOptions.RunContinuationsAsynchronously);
+            private readonly int _removeThreshold;
+
+            private int _addCount;
+            private int _removeCount;
+
+            public ThresholdCallback(int addThreshold, int removeThreshold = 0)
+            {
+                _addThreshold = addThreshold;
+                _removeThreshold = removeThreshold;
+
+                SetIfAtTarget(_addCompletionSource, addThreshold, 0);
```

Comment:
I don't think this matters from a functionality standpoint, but this felt flipped while I was reading this (i.e. it should be `SetIfAtTarget(_addCompletionSource, 0, addThreshold`). It only applies if they're both 0 so it doesn't matter, but this states that the threshold is the value instead of the target (which is how this is represented in `AfterAdd` and `AfterRemove`). Maybe I'm misinterpreting it, but it seemed off when I was reading this.